[BladerX11] iP

BladerX11 opened the pull request:
# Tasker
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))
DukePro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~FAST~ _BLASINGLY_ FAST to use

All you need to do is,

1. double-click it.
2. add your tasks.
3. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(Ui.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

BladerX11 committed:
Refactor: Change product name to Tasker

BladerX11 committed:
Feature: Add skeletal greeting and exit

BladerX11 committed:
Refactor: Add padding to responses

BladerX11 committed:
Feature: add support for echo

BladerX11 committed:
Feature: Add TaskList class to track tasks

BladerX11 committed:
Refactor: Move classes into packages

BladerX11 committed:
Refactor: Set TaskList class to public

BladerX11 committed:
Docs: Add Javadoc to classes

BladerX11 committed:
Feature: Add ability to add tasks and list them

BladerX11 committed:
Chore: Add class files to .gitignore

BladerX11 committed:
Feature: Add Task class to represent tasks

BladerX11 committed:
Refactor: Change list display

BladerX11 committed:
Feature: Add ability to get Task from TaskList

BladerX11 committed:
Feature: Add marking and unmarking tasks as done

BladerX11 committed:
Feature: Add ability to add Todos

BladerX11 committed:
Refactor: Change Tasklist getTask name

BladerX11 committed:
Feature: Show total task count when adding

BladerX11 committed:
Refactor: Use multi-line strings and String.format

BladerX11 committed:
Docs: Update Javadocs

BladerX11 committed:
Feature: Add tasks with deadline

BladerX11 committed:
Feature: Add ability to add Events to task list

BladerX11 committed:
Test: Compile all java files for tests

BladerX11 committed:
Test: Add input and expected output

BladerX11 committed:
Refactor: adjust response alignment

BladerX11 committed:
Refactor: use enums for valid commands

BladerX11 committed:
Refactor: use StringBuilder for response

BladerX11 committed:
Feature: add listting valid commands

BladerX11 committed:
Feature: list valid commands when invalid one is used

BladerX11 committed:
Feature: check empty todo description

BladerX11 committed:
Docs: update javadoc

BladerX11 committed:
Feature: Detect unknown commands

BladerX11 committed:
Refactor: Move MarkingCommand to IndexCommand

BladerX11 committed:
Refactor: Report task list size after adding tasks

BladerX11 committed:
Refactor: Change command to enum before comparing

BladerX11 committed:
Refactor: use switch statement instead of if-else

BladerX11 committed:
Feature: Add ability to delete tasks

BladerX11 committed:
Feature: Check invalid deadline and event command

BladerX11 committed:
Feature: Check validity of mark, unmark and delete

BladerX11 committed:
Refactor: Add command for bye to handle exiting

BladerX11 committed:
Refactor: use isDone setter for marking tasks

BladerX11 committed:
Refactor: Fix java code style

BladerX11 committed:
Refactor: Add label property to Task class

BladerX11 committed:
Refactor: rename Parser parse to parseCommand

BladerX11 committed:
Feature: Add ability to create task with isDone set

BladerX11 committed:
Feature: Add loading tasks from file

BladerX11 committed:
Feature: Add saving tasks to storage

BladerX11 committed:
Merge branch 'branch-Level-7'

BladerX11 committed:
Refactor: Create method to get index task description

BladerX11 committed:
Refactor: Make IndexCommand fields private

BladerX11 committed:
Feature: Recognize dates in Deadline tasks

BladerX11 committed:
Refactor: Wrap lines if it exceeds separator

BladerX11 committed:
Feature: Use LocalDateTime for Event tasks

BladerX11 committed:
Refactor: Move Parser to tasker package

BladerX11 committed:
Test: Clean up data after testing

BladerX11 committed:
Merge branch 'branch-Level-8' into HEAD

BladerX11 committed:
Refactor: accept TaskList for save in Storage

BladerX11 committed:
Refactor: Use Storage for loading in TaskList

BladerX11 committed:
Refactor: Use enum for TaskType

BladerX11 committed:
Refactor: Extract UI logic to Ui class

BladerX11 committed:
Feature: Create new task list if loading fails

BladerX11 committed:
Merge branch 'add-gradle-support'

BladerX11 committed:
Chore: remove .class files from gitignore

BladerX11 committed:
Build: Set correct values for gradle file

BladerX11 committed:
Refactor: Add equals and hashcode for tasks

BladerX11 committed:
Refactor: add equals and hashCode for commands

BladerX11 committed:
Test: Add tests for Parser

BladerX11 committed:
Refactor: Change Event invalid date format

BladerX11 committed:
Test: Add Todo tests

BladerX11 committed:
Build: Add jar classifier to indicate fat jar

BladerX11 committed:
Docs: Update task JavaDocs

BladerX11 committed:
Merge branch 'branch-A-JavaDoc'

BladerX11 committed:
Refactor: Fix ParserTest spacing

BladerX11 committed:
Merge branch 'branch-A-CodingStandard'

BladerX11 committed:
Feature: Add ability to find task with description

BladerX11 committed:
Merge branch 'branch-Level-9'

BladerX11 committed:
Fix: Fix AddCommand hashCode method

BladerX11 committed:
Fix: Fix restoring tasks from storage

BladerX11 committed:
Build: Add Checkstyle

BladerX11 committed:
Refactor: Fix styling

BladerX11 committed:
Merge branch 'branch-A-CheckStyle'

BladerX11 committed:
Build: Add JavaFX

BladerX11 committed:
Feature: Add GUI to application

BladerX11 committed:
Fix: Restore functionality to GUI

BladerX11 committed:
Refactor: Remove error printing and edit JavaDoc

BladerX11 committed:
Test: Remove I/O redirection automated testing

BladerX11 committed:
Merge branch 'branch-Level-10'

andong0909 commented on src/main/java/tasker/Parser.java:
```diff
@@ -0,0 +1,214 @@
+    public static Command parseCommand(String command) throws TaskerException {
+        String[] cmdParts = command.split(" ", 2);
+        CommandType mainPart;
+        Command toRun = null;
```
Should the naming be like this? Maybe consider changing to a noun?

andong0909 commented on src/main/java/tasker/Parser.java:
```diff
@@ -0,0 +1,214 @@
+                    %s""", cmdParts[0], CommandType.listCommands()));
+        }
+
+        switch (mainPart) {
```
Nested switch/case block which can be hard to read & understand, especially with fall through cases. May you can consider refactoring / comments on each case on what the code does?

andong0909 commented on src/main/java/tasker/Parser.java:
Consider using separate function for repeated logics instead of using nested switch/case blocks

andong0909 commented on src/main/java/tasker/command/ByeCommand.java:
```diff
@@ -0,0 +1,22 @@
+ */
+public class ByeCommand extends Command {
+    /**
+     * Exits the program.
```
Does it really exit the programme? Maybe can be more precise!

andong0909 reviewed (commented):
Good job! Generally clear structure and functional code!
Maybe some pointers for improvement you can consider:
1. Consider abstracting away the messages to be printed Ui class and simply call for the function instead.
2. Consider having more specific JavaDoc comments for precision and easier understanding!
🥇 🥇

minhkiwi14 commented on src/main/java/tasker/Parser.java:
```diff
@@ -0,0 +1,214 @@
+     * @return The command to handle the user input.
+     * @throws TaskerException If there is an error with the command
+     */
+    public static Command parseCommand(String command) throws TaskerException {
```
I like the use of verb phrases like this.

minhkiwi14 commented on src/main/java/tasker/Parser.java:
```diff
@@ -0,0 +1,214 @@
+        case DEADLINE:
+        case EVENT:
+        case TODO:
+            Task toAdd = null;
```
Perhaps a noun is more appropriate?

minhkiwi14 commented on src/main/java/tasker/Parser.java:
```diff
@@ -0,0 +1,214 @@
+
+    public static Task parseStorage(String line) throws TaskerException {
+        String[] parts = line.split("\\|");
+        TaskerException incorrectFormat = new TaskerException("Incorrect storage format");
```
Maybe consider changing to `incorrectFormatException` to increase the interpretability?

minhkiwi14 commented on src/main/java/tasker/Tasker.java:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Main class for project.
+ */
+public class Tasker {
```
Is it better to rename it `TaskManager` or similarly (I'm not sure if Tasker is a valid English word)?

minhkiwi14 reviewed (commented):
LGTM. Just a few nits to fix.

BladerX11 committed:
Refactor: Use nouns for variable names in Parser

BladerX11 committed:
Refactor: Extract parts of parseCommand to methods

Currently, the parseCommand method is overly complex due to deep nesting
and multiple responsibilities, which makes it hard to read, test, and
maintain.

It needs to change because the method violates code quality guidelines by
not adhering to the single responsibility principle and by containing
deeply nested control structures, leading to potential bugs and
difficulties in future modifications.

Let's
* Extract helper methods for task commands, index-based commands, and
  other commands
* Replace nested switch cases with simpler control flows and guard
  clauses to validate inputs.

This approach improves readability, maintainability, and adherence to
best practices, making it easier to manage and extend the command
parsing functionality.

BladerX11 committed:
Refactor: Split parseStorage into multiple methods

The parseStorage method currently handles all task parsing logic in a
single block, resulting in reduced readability.

This monolithic structure complicates maintenance and testing.

Refactor the method by extracting helper functions to handle
task-specific parsing.

This isolates the parsing logic into dedicated helper methods improves
code clarity.

BladerX11 committed:
Merge pull request #1 from BladerX11/branch-A-CodeQuality

Branch a code quality

BladerX11 committed:
Refactor: Use streams for list processing

The TaskList class uses manual iteration with for loops to load tasks
from Storage and filter them, while the Storage class employs loops to
serialize and save tasks to disk. This approach relies on mutable
intermediate collections and imperative-style operations.

Explicit loops and manual iteration add boilerplate code, reduce
readability, and increase the risk of off-by-one errors. Streams
provide a more declarative, concise, and maintainable way to process
collections.

Let's
* Refactor TaskList to use Java Streams for loading and filtering tasks,
replacing loops with `stream()`.
* Replace intermediate list operations with method references and stream pipelines.

Streams eliminate mutable state, reduce side effects, and clarify data
flow. Collectors like `toList()` provide immutable results by default,
improving thread safety.

BladerX11 committed:
Merge branch 'master' into branch-A-Streams

BladerX11 committed:
Merge pull request #2 from BladerX11/branch-A-Streams

Branch a streams

BladerX11 committed:
Refactor: Use method references for stream methods

BladerX11 committed:
Refactor: Fix naming on Parser method

BladerX11 committed:
Refactor: Use assertions to enusre assumptions

The Parser class initializes a variable to null by default, which can
result in unintended null states if the variable is not properly set.

This situation needs to change because leaving the variable null may
lead to unexpected NullPointerExceptions during runtime and complicate
debugging.

Add assertions to verify that the variable is no longer null after its
initialization, ensuring that the object state transitions as expected.

Using assertions enables early detection of potential errors and helps
maintain robustness in the parser logic by enforcing correct state
changes.

BladerX11 committed:
Merge branch 'master' into branch-A-Assertions

BladerX11 committed:
Merge pull request #3 from BladerX11/branch-A-Assertions

Branch a assertions

BladerX11 committed:
Refactor: Split parsing add command into methods

BladerX11 committed:
Feature: Add tasks with fixed duration to complete

BladerX11 committed:
Feature: Parse fixed duration tasks

BladerX11 committed:
Test: Add tests for parsing fixed duration task

BladerX11 committed:
Merge branch 'branch-B-FixedDurationTasks'

BladerX11 committed:
Refactor: Abstract storage transformation to TaskList

BladerX11 committed:
Refactor: Rename DialogBox to reflect Tasker

BladerX11 committed:
Refactor: Use smallest scope for DialogBox variable

BladerX11 committed:
Test: Fix test for missing description for tasks

BladerX11 committed:
Refactor: Reduce command parts passing in Parser

BladerX11 committed:
Feature: Add name to title bar

BladerX11 committed:
Fix: Load fix duration marking from storage

BladerX11 committed:
Docs: Add screenshot of app

BladerX11 committed:
Fix: Prevent error message truncation

BladerX11 committed:
Docs: Add features and usage examples

BladerX11 committed:
Merge branch 'branch-A-UserGuide'

BladerX11 committed:
Refactor: Remove unused imports

BladerX11 committed:
Fix: Add spacing before results of find

BladerX11 committed:
Docs: Specify description comes before argument separators

BladerX11 committed:
Refactor: Remove white space in command components

BladerX11 committed:
Refactor: Simplify storage parsing

BladerX11 committed:
Fix: Show H behind time in output

BladerX11 committed:
Docs: Clarify parameter pupose and examples

BladerX11 committed:
Fix: Disallow `|` in description for affecting save

BladerX11 committed:
Docs: Add notice about ignoring whitespace

BladerX11 committed:
Fix: Fix DialogBox failure typo

BladerX11 committed:
Fix: Show intended text even when DialogBox fails

BladerX11 committed:
Fix: Show correct invalid index in response

BladerX11 committed:
Feature: Use tasker to show show errors

BladerX11 committed:
Docs: Add notes on repeated marking and unmarking

BladerX11 committed:
Docs: Improve clarity of docs

BladerX11 committed:
Fix: Prevent event start from coming after end

BladerX11 committed:
Docs: indicate events and deadlines can be past

BladerX11 committed:
Fix: Prevent negative fixed duration hrs and mins

BladerX11 committed:
Fix: Check if index >= 0 for valid index

BladerX11 committed:
Feature: Add initial greeting

BladerX11 committed:
Docs: Fix grammar

BladerX11 committed:
Docs: Update screenshot

BladerX11 committed:
Docs: Update README

BladerX11 committed:
Docs: Add screenshot to README

BladerX11 committed:
Docs: Format README
